specification/dartLangSpec.tex

```diff
@@ -16484,160 +16768,282 @@ \subsection{Exports}
-for each public top-level declaration $D$ with name $n$ in $L$ maps $n$ to $D$.
+In the first step we compute the namespace provided by each exported library:
 For each $i \in 1 .. m$,
-let $\Namespace{\metavar{export}, i}$ be the exported namespace of $L_i$,
-and let \Namespace{i} be a namespace obtained from
-$\Namespace{\metavar{export}, i}$ by the narrowing process described below.
+\NamespaceName{\metavar{exported},i} is
+the namespace obtained from applying
+the namespace combinators of $E_i$ to
+the exported namespace of $L_i$
+(\ref{namespaceCombinators}),
+and removing each binding of a name $n$ such that
+\NamespaceName{\metavar{public}} is defined at $n'$,
+and $n$ and $n'$ have the same basename.
+\commentary{So local declarations shadow re-exported ones.}
 
 \LMHash{}%
-The narrowing process that yields
-\Namespace{i} from $\Namespace{\metavar{export}, i}$
-consists of two steps.
-Each of them restricts the domain of the namespace to a subset
-(\commentary{%
-that is, it deletes the binding from a key to a value,
-for certain keys%
-}).
+The
+\IndexCustom{namespace re-exported from}{library!namespace re-exported from}
+$L_i$ is \NamespaceName{\metavar{exported},i}.
 
-{ %% Local scope: abbreviate narrowing namespaces.
+\LMHash{}%
+In the second step we compute the exported namespace of $L$.
+Let \NamespaceName{\metavar{merged}} be the result of applying
+conflict merging
+(\ref{conflictMergingOfNamespaces})
+to \List{\metavar{NS}}{\metavar{one},1}{\metavar{one},m}.
+A compile-time error occurs if any name in
+\NamespaceName{\metavar{merged}}
+is conflicted
+(\ref{conflictMergingOfNamespaces}).
+
+\rationale{%
+This rule is more strict than the corresponding rule for imports:
+When two imported declarations have a name clash it is only an error to
+\emph{use} the conflicted name, it is not an error that the name clash exists.
+But with exported names it is an error that the name clash exists.
+The reason for this difference is that
+the conflict could silently break importers of the current library $L$,
+if we were to use the same approach for exports as for imports:
+If a library $L'$ imports $L$
+and uses a name $n$ which is re-exported by $L$ from $L_n$
+then the addition of a declaration named $n$ to some other re-exported library
+will make the use of $n$ in $L'$ an error,
+and the maintainers of $L'$ may not be in a position to change $L$.%
+}
 
-% The sequence of narrowed name spaces of export i.
-\def\NSNi#1{\Namespace{\metavar{narrowing},i,{#1}}}
+\LMHash{}%
+The
+\IndexCustom{exported namespace}{library!exported namespace of}
+of $L$ is
+$\NamespaceName{\metavar{public}} \cup \NamespaceName{\metavar{merged}}$.
 
-% The final narrowed name space of a given export. We don't want to
-% mention the number of hide/show clauses when the export varies.
-\def\NSN#1{\Namespace{\metavar{narrowing},{#1}}}
+\LMHash{}%
+We say that a name is \Index{exported by a library}
+if the name is in the library's exported namespace.
```

Comment:
(Do we say that a name is in a namespace if a declaration with that name is in the namespace)?

Reply:
Yes, search for `a name $n$ \Index{is in}`.